pkg/lib/indexer/indexer.go

```diff
@@ -570,19 +586,83 @@ func (i ImageIndexer) ExportFromIndex(request ExportFromIndexRequest) error {
 }
 
 type bundleDirPrefix struct {
-	pkgName, bundleVersion string
+	pkgName, name, bundleVersion, replaces string
+	skips                                  map[string]struct{}
 }
 
 func getBundlesToExport(dbQuerier pregistry.Query, packages []string) (map[string]bundleDirPrefix, error) {
 	bundleMap := make(map[string]bundleDirPrefix)
-
+	graphLoader := &sqlite.SQLGraphLoader{
+		Querier: dbQuerier,
+	}
 	for _, packageName := range packages {
-		bundlesForPackage, err := dbQuerier.GetBundlesForPackage(context.TODO(), packageName)
+		bundleGraph, err := graphLoader.Generate(packageName)
 		if err != nil {
 			return nil, err
 		}
-		for k, _ := range bundlesForPackage {
-			bundleMap[k.BundlePath] = bundleDirPrefix{pkgName: packageName, bundleVersion: k.Version}
+		versionMap := make(map[string]string)
+		for _, channel := range bundleGraph.Channels {
+			var replaceDepth map[string]int64
+			for node, replaceNodes := range channel.Nodes {
+				var replaces string
+				var skips map[string]struct{}
+				if bundleInfo, ok := bundleMap[node.BundlePath]; ok {
+					replaces = bundleInfo.replaces
+					skips = bundleInfo.skips
+				}
+				if skips == nil {
+					skips = map[string]struct{}{}
+				}
+				skips[replaces] = struct{}{}
+				for replace := range replaceNodes {
+					if replace.CsvName == replaces {
+						continue
+					}
+					if _, ok := channel.Nodes[replace]; !ok {
+						//synthetic entry
+						skips[replace.CsvName] = struct{}{}
+						continue
+					}
+					if len(replaces) == 0 {
+						replaces = replace.CsvName
+						versionMap[replaces] = replace.Version
+						continue
+					}
+
+					if replaceDepth == nil {
+						replaceDepth, err = dbQuerier.GetBundleReplacesDepth(context.TODO(), packageName, node.CsvName)
+					}
+					if err == nil && replaceDepth[replaces] != replaceDepth[replace.CsvName] {
+						if replaceDepth[replaces] < replaceDepth[replace.CsvName] {
+							replaces = replace.CsvName
+							versionMap[replaces] = replace.Version
+						} else {
+							skips[replace.CsvName] = struct{}{}
+						}
+						continue
+					}
+
+					replacesSemver, replacesErr := semver.Make(versionMap[replaces])
+					newSemver, newErr := semver.Make(replace.Version)
+					if replacesErr == nil && newErr == nil && !replacesSemver.EQ(newSemver) {
+						if replacesSemver.LT(newSemver) {
+							replaces = replace.CsvName
+							versionMap[replaces] = replace.Version
+						} else {
+							skips[replace.CsvName] = struct{}{}
+						}
+						continue
+					}
+					return nil, fmt.Errorf("multiple replaces on package %s, bundle version %s: %s and %s", packageName, node.CsvName, replaces, replace.CsvName)
```

Comment:
It looks like no tests are covering this section (https://codecov.io/gh/operator-framework/operator-registry/pull/504/diff#D1-632).